### PR TITLE
OSCDDataModule: add random crop augmentation

### PR DESCRIPTION
**Pull request by iejMac**

### CODE REVIEW

**Comment by isaaccorley** [comment]
This seems to reinvent the wheel. We should always choose to use a transform that's available in either torchvision or kornia. RandomCrop can be found in both. Also if you are cropping the image you will need to crop the change mask as well. 

Take a look at torchgeo.transforms.AugmentationSequential. You can use this along with kornia's RandomCrop to crop both image/masks in the same way.

**Comment by isaaccorley** [comment]
You are doing random cropping for the val and test set which means you will not use the same set of data when computing metrics when benchmarking. You can use kornia.contrib.ExtractTensorPatches and kornia.contrib.CombineTensorPatches to chip up each image and feed through the model and then combine the patches and predicted change masks to compute metrics.

**Comment by adamjstewart** [comment]
Or we could not use random cropping during val/test...

**Comment by isaaccorley** [comment]
Yup that's what I meant. The random cropping should only be used during training but for val and test we still need to chip up the imagery because of the large variance in height/width.

**Comment by calebrob6** [comment]
Few things:
- Cropping of some sort is needed at the datasets level here as the samples are all different sizes. A dataloader will not be able to combine them into a batch otherwise.
- Random cropping is OK during validation, although it will give noisier performance estimates epoch over epoch.
- The test set could just use a batch size of 1 to ignore this issue (same with the validation set). There are only 14 training samples (that will get further split into train/val) and 10 test samples.
- @isaaccorley can you repost the ExtractTensorPatches and CombineTensorPatches snippet you have to chip up a larger image here?
  - This is not a perfect solution because if a model uses zero-padding it will introduce potential edge effects over the chips.

**Comment by calebrob6** [comment]
To summarize @iejMac, I think the easiest thing to do is change batch size in the val and test dataloaders to 1, then only do the random cropping on the train dataset.

@isaaccorley @adamjstewart any ideas on how to get multiple random crops for a single input to the dataloader?

**Comment by isaaccorley** [comment]
We need to be careful using `AugmentationSequential` for change detection here. Because `sample["image"]` is of shape `(2, c, h, w)` it's being treated as a batch so you'll get two different random crops. Easy fix is to set `same_on_batch=True`. You'll also need to repeat the mask (I use einops here but you can use `torch.repeat` if you're more comfortable with that) so that it has the same temporal dim as the images and then just grab the first mask afterwards.

```python
crop_size = (128, 128)
rcrop = K.AugmentationSequential(
    K.RandomCrop(crop_size),
    data_keys=["input", "mask"],
    same_on_batch=True
)

def transform(sample):
    sample["mask"] = repeat(sample["mask"], "h w -> t h w", t=2)
    sample["image"], sample["mask"] = rcrop(sample["image"], sample["mask"].to(torch.float))
    sample["mask"] = sample["mask"].squeeze()[0].to(torch.long)
    return sample
```

Example with `same_on_batch=False`:
image

Example with `same_on_batch=True`:
image

**Comment by isaaccorley** [comment]
> To summarize @iejMac, I think the easiest thing to do is change batch size in the val and test dataloaders to 1, then only do the random cropping on the train dataset.
> 
> @isaaccorley @adamjstewart any ideas on how to get multiple random crops for a single input to the dataloader?

I don't think it would be difficult to write a custom transform that just repeats `K.RandomCrop` `n` times and then stacks the resulting images and masks. The dataloader will likely load them incorrectly as a batch of `(t, c, h, w)` images so will have to flatten dim=1 into the batch dim.

**Comment by isaaccorley** [comment]
Here's an example of extracting patches. You'll have to use `kornia.contrib.CombineTensorPatches` but that will need to go somewhere in the trainer.

```python
from typing import Tuple

import torch
import torch.nn as nn
import torchvision.transforms as T
from kornia.contrib import ExtractTensorPatches
from torchvision.utils import make_grid

from torchgeo.datasets import OSCD

root = "path/to/root"
ds = OSCD(root, split="train")
sample = ds[0]
image = sample["image"][:, [3, 2, 1]] # (2, 3, 799, 785)
h, w = image.shape[-2:] # (799, 785)
patch_size = (128, 128)


class PadToDivisible(nn.Module):
    """ Pad an input tensor to be divisible by a h, w. """
    def __init__(self, size: Tuple[int, int]) -> None:
        super().__init__()
        self.h, self.w = size

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        xh, xw = x.shape[-2:]
        pad_h, pad_w = self.h - (xh %  self.h), self.w - (xw % self.w)
        x = F.pad(x, (0, pad_w, 0, pad_h))
        return x

# Extract non overlapping patches, zero pad uneven edges
transforms = nn.Sequential(
    PadToDivisible(patch_size),
    ExtractTensorPatches(window_size=patch_size, stride=patch_size)
)

patches = transform(image)
patches.shape # (2, 49, 3, 128, 128)
grid = make_grid(patches[0, ...], nrow=7)
T.ToPILImage()(grid)
```

image

File: torchgeo/datasets/oscd.py [hunk 1]
```diff
@@ -414,22 +422,33 @@ def setup(self, stage: Optional[str] = None) -> None:
...
         dataset = OSCD(
-            self.root_dir, split="train", bands=self.bands, transforms=transforms
+            self.root_dir, split="train", bands=self.bands, transforms=train_transforms
         )
 
         if self.val_split_pct > 0.0:
             self.train_dataset, self.val_dataset, _ = dataset_split(
                 dataset, val_pct=self.val_split_pct, test_pct=0.0
             )
+            self.val_dataset.transforms = test_transforms
```
Comment (iejMac) [inline torchgeo/datasets/oscd.py:445]: Unsure why this isn't working. The val_dataset and train_dataset are split from the same dataset and I need to override the val_datasets transform somehow.

Comment (isaaccorley) [reply torchgeo/datasets/oscd.py:445]: If you look at the return value of dataset_split you will see that it returns a torch.utils.data.Subset object. If you look up what a Subset object is you will find that it has an attribute Subset.dataset which holds the original dataset and samples indices from it. You can override the transforms via self.val_dataset.dataset.transforms

**Comment by iejMac** [comment]
This is ready for review / merge

File: torchgeo/datasets/oscd.py [hunk 2]
```diff
@@ -391,14 +397,21 @@ def __init__(
...
+        self.rcrop = K.AugmentationSequential(
+            K.RandomCrop(crop_size[1:]), data_keys=["input", "mask"], same_on_batch=True
+        )
 
     def preprocess(self, sample: Dict[str, Any]) -> Dict[str, Any]:
         """Transform a single sample from the Dataset."""
         sample["image"] = sample["image"].float()
+        sample["mask"] = sample["mask"].float()
         sample["image"] = self.norm(sample["image"])
         sample["image"] = torch.clamp(  # type: ignore[attr-defined]
```
Comment (calebrob6) [inline torchgeo/datasets/oscd.py:409]: This clamp line should be removed. You are standardizing per channel, which means the data will have 0 mean and 1 stdev -- if you clamp then you are getting rid of > 1/2 of your data.

File: torchgeo/datasets/oscd.py [hunk 3]
```diff
@@ -375,13 +379,15 @@ def __init__(
             batch_size: The batch size to use in all created DataLoaders
             num_workers: The number of workers to use in all created DataLoaders
             val_split_pct: What percentage of the dataset to use as a validation set
+            crop_size: Size of random crop from image and mask (n, height, width)
```
Comment (calebrob6) [inline torchgeo/datasets/oscd.py:382]: I think you should split crop_size up into something like `num_patches_per_tile` and `patch_size`, currently it is unintuitive (maybe just to me!) to have the number of samples in something called `crop_size`

File: torchgeo/datasets/oscd.py [hunk 4]
```diff
@@ -362,9 +365,10 @@ def __init__(
         self,
         root_dir: str,
         bands: str = "all",
-        batch_size: int = 64,
+        batch_size: int = 32,
```
Comment (calebrob6) [inline torchgeo/datasets/oscd.py:368]: This maybeee should be called `train_batch_size` with documentation in the constructor explaining that the test and val dataloader will use batch size of 1.

**Review by calebrob6 — COMMENTED** [review]
Nice job! I verified that the train_dataloader is returning what we expect.

Can you change the docstring on the DataModule to something other than "Uses the train/val/test splits from the dataset." We use the train/test split from the dataset and further split the train split into train/val.

File: torchgeo/datasets/oscd.py (hunk 3 above)
Comment (iejMac) [reply torchgeo/datasets/oscd.py:383]: Yeah I was also unsure of this, just wanted to keep the # of args down but maybe it would be worth it to split it up in this case.

**Review by calebrob6 — APPROVED** [review]
